fml/message_loop_impl.cc

```diff
@@ -0,0 +1,96 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/fml/message_loop_impl.h"
+#include "lib/ftl/build_config.h"
+
+#if OS_MACOSX
+
+#include "flutter/fml/platform/darwin/message_loop_darwin.h"
+using PlatformMessageLoopImpl = fml::MessageLoopDarwin;
+
+#else
+
+#error This platform does not have a message loop implementation.
+
+#endif
+
+namespace fml {
+
+ftl::RefPtr<MessageLoopImpl> MessageLoopImpl::Create() {
+  return ftl::MakeRefCounted<::PlatformMessageLoopImpl>();
+}
+
+MessageLoopImpl::MessageLoopImpl() : order_(0) {}
+
+MessageLoopImpl::~MessageLoopImpl() = default;
+
+void MessageLoopImpl::PostTask(ftl::Closure task, ftl::TimePoint target_time) {
+  if (task == nullptr) {
+    return;
+  }
+
+  WakeUp(RegisterTaskAndGetNextWake(task, target_time));
+}
+
+void MessageLoopImpl::RunExpiredTasksNow() {
+  WakeUp(RunExpiredTasksAndGetNextWake());
+}
+
+void MessageLoopImpl::SetTaskObserver(MessageLoop::TaskObserver observer) {
+  ftl::MutexLocker lock(&task_observer_mutex_);
+  task_observer_ = observer;
+}
+
+ftl::TimePoint MessageLoopImpl::RegisterTaskAndGetNextWake(
+    ftl::Closure task,
+    ftl::TimePoint target_time) {
+  FTL_DCHECK(task != nullptr);
+  ftl::MutexLocker lock(&delayed_tasks_mutex_);
+  delayed_tasks_.push({++order_, std::move(task), target_time});
+  return delayed_tasks_.top().target_time;
+}
+
+ftl::TimePoint MessageLoopImpl::RunExpiredTasksAndGetNextWake() {
+  std::vector<ftl::Closure> invocations;
+
+  {
+    ftl::MutexLocker lock(&delayed_tasks_mutex_);
+
+    if (delayed_tasks_.empty()) {
+      return ftl::TimePoint::Max();
+    }
+
+    auto now = ftl::TimePoint::Now();
+    while (!delayed_tasks_.empty()) {
```

Comment:
Rather than doing all this work under delayed_tasks_mutex_, the design we've used in other places is to have the lock protect an std::vector.  Then, under the lock you can just swap the vector into a local variable.

After releasing the lock, you can dump the vector into a priority queue that's held on the loop's thread.  You can then pull from the priority queue without having to hold the delayed_tasks_mutex_